### ISSUE TITLE
STATUS_BAD_NETWORK_NAME: 0xc00000cc

### ISSUE THREAD

**Issue by i3uiw** [jborean93/smbprotocol#81]
Good day folks.  Looking for some assistance.  I'm getting the above error upon a write to a DFS share.  The piece of code is working as I do get files/etc on the share but the script is long running (should take up to 5 hours).  Fails at around the 35 minute mark.  Files are around the 100 to 200 MB size.

Any advice on how to fix?

Overview of code/environment:
smbprotocol v1.4.0 (Python 3.8.3)
importing smbclient module only
ClientConfig using defaults (except user/password)
file contents is the text output from a SSH session (mode="w", share_access="r", encoding='utf-16')
file handler closed after write

```
Traceback (most recent call last):
	File "\Python\Python38\lib\site-packages\smbclient\_pool.py", line 302, in get_smb_tree
		tree.connect(require_secure_negotiate=client_config.require_secure_negotiate)
	File "\Python\Python38\lib\site-packages\smbprotocol\tree.py", line 245, in connect
		response = self.session.connection.receive(request)
	File "\Python\Python38\lib\site-packages\smbprotocol\connection.py", line 617, in wrapped
		return func(self, *args, **kwargs)
	File "\Python\Python38\lib\site-packages\smbprotocol\connection.py", line 931, in receive
		raise SMBResponseException(response)
	smbprotocol.exceptions.BadNetworkName: Received unexpected status from the server: The specified share name cannot be found on the remote server. (3221225676) STATUS_BAD_NETWORK_NAME: 0xc00000cc
```

**Comment by jborean93** [comment]
It's really hard to tell what may be happening here, DFS is a very complex protocol to try and implement and I'm not surprised there are bugs in it. Here is the general workflow when you try to connect to a DFS share, for example we try to open the file at  `\\domain.com\share\file.txt`

1. We check the DFS referral cache to see if `\\domain.com\share` matches any cached referals
2. Because this is the first request there will be no cache so we just connect to `\\domain.com\share`
3. `domain.com` will probably resolve to a domain controller and it returns `STATUS_BAD_NETWORK_NAME` because it does not have a share named `share`
4. We try to open `\\domain.com\IPC$`
5. We send a DFS request to `\\domain.com\IPC$` to get the referral for `\domain.com\share` which would return something like `\\server1\share`
6. We cache the referral response with a timeout value so subsequent requests to that DFS share know exactly what host to talk to
7. We then open a connection to `\\server1\share` and continue on as normal

The only time `get_smb_tree` raises `BadNetworkName` is if we tried to connect to `IPC$` share, if that fails we have nothing to send the referral request to so an exception is the only option here.

Would you be able to answer the following

* Is your DFS share based on a Windows host or some other SMB implementation like Samba, FreeNAS, etc
* Are the shares you are connecting to all on the same DFS paths or do they vary
* Are you setting the `ClientConfig(domain_controller='something')`?
* Is there a way you can try and narrow down this failure a bit more, i.e. does it always happen to the same DFS share or is it time based
* Anything else that you think might help narrow this down

**Comment by i3uiw** [comment]
Answers:

- This is in a Windows environment for authentication BUT uses Netapp as the back end storage w/ Infoblox DNS.
- The paths only change in file name only.
- I am not setting domain_controller var.  Should I be?
- Best I have is it is time based failure.  Like a worker thread has died and the client cannot connect to the worker?  How can a worker get re-started?  Or set the cache to stale and force a refresh from DNS?
- Some oddities have come up in the SSH feed that is being saved in that the character set causes problems.  Hence the UTF-16 vs. std UTF-8. I did try at one point to perform a binary save (mode="wb") but the result was messy (I don't remember the error codes).  I'll see about re-producing this error next week and report back with results but don't think this is an issue with the smbprotocal module.
- Last oddity is that when the error happens, Python eventually reaches max recursion on the error as it is trying to process the error results.   IE: error causes error which causes error which causes error (repeating).  All with the same traceback of STATUS_BAD_NETWORK_NAME: 0xc00000cc.

I'm looking into other ways to 'fool' the module over the next few days.
- Breakup the stream of data to be stored into several different write/appends to the same file.  Probably on line breaks or some other value akin to block size lineup.
- Changing the script to store locally and then transfer it via fs objects.
- Zip the output in memory and then save to the share.
- Remove DFS from the mix (not ideal)

**Comment by jborean93** [comment]
> The paths only change in file name only.

Ok, just to confirm you are trying to do stuff with files/folder in `\\server\share\x` where anything in `x` or beyond is variable? IIRC DFS can have link targets where paths inside a DFS namespace can then refer to another DFS target, i.e. `\\server\share\folder\link` should point to `\\other-server\share`. I can't remember actually testing this so there's a chance.

> I am not setting domain_controller var. Should I be?

It's up to you, the benefit of setting `domain_controller` is that it will cause `smbclient` to send a domain referral request to `\\domain_controller\IPC$` to get back a list of valid domain names (usually the NETBIOS and DNS name, `DOMAIN` and `domain.com`). If you then try to connect to `\\DOMAIN` or `\\domain.com` then a dc referral request is set to the domain controller to get a prioritised list of domain controllers that root referrals should be sent to. The benefit of this approach is you can now connect to a host with either the `\\NETBIOS` or `\\dns.com` names rather than just the typical `dns.com` name and relying on DNS to resolve that to a domain controller. It could help as the referrals happen before trying to connect to `\\domain\share` as it will check to see if `\\domain\share` is a root referral itself rather than relying on the `BadNetworkName` fallback.

> Best I have is it is time based failure. Like a worker thread has died and the client cannot connect to the worker?

If the worker thread has died you will see a hang as documented in other issues https://github.com/jborean93/smbprotocol/issues/78. The fact that you are getting the status code back means the worker thread is actually running and correctly returning the response.

> Some oddities have come up in the SSH feed that is being saved in that the character set causes problems

I'm not 100% sure how this relates to the problem. You tell SMB to write text or bytes and the encoding is up to you and your environment. If you wish to write text then that's converted to the bytes you specify when you've opened the file. If you read/write in binary mode it's entirely up to you to write the correct bytes.

> Last oddity is that when the error happens, Python eventually reaches max recursion on the error as it is trying to process the error results

That would indicate this is running in a perpetual loop https://github.com/jborean93/smbprotocol/blob/master/smbclient/_pool.py#L301-L309 but I don't fully see how. It:

* Attempts to connect to `\\server\share`
* That fails so we connect to `\\server\IPC$`
* If that fails then it matches the `if path == ipc_path:` check and the `BadNetworkName` error is raised
* If that doesn't fail then we send the DFS referral request to find out what host `\\server\share` is located on

What would be good to see is a wireshark capture of the failure. That will show what path we tried to connect to and the subsequent fallback behaviour we tried to do to connect to that DFS path. This might contain sensitive information so I've mostly just interested in the final stage that failed, i.e. the `TREE_CONNECT` and subsequent DFS requests to IPC$ that might be raising the exception.

**Comment by i3uiw** [comment]
Yeah.  I hated having to send out a generic 'It's failing!  Fix it!' issue.  Wireshark data (after cleaned) forthcoming.
In pseudo code, here is the loop:
```
import smbclient
path = r"\\server\share\folder\folder2\folder3"
file_name_list = ['orange', 'apple', 'pear', 'banana', 'kiwi', 'apple', 'banana']
for file_name in file_name_list:
    smbclient.ClientConfig(username=user, password=pwd)
    with smbclient.open_file(path + "\\" + file_name + ".txt", mode="w", share_access="r", encoding="utf-16") as fh:
       fh.write(some_data)
       fh.close()
```

**Comment by i3uiw** [comment]
Ok.  Wireshark info (sanitized).  Not knowing what to pass/sanitize, here is some initial detail.  I have the packets captured so can grab whatever else is needed.  Sanitized sections will be encompassed by "<" and ">".  This run failed at the 34 minute mark.

1. Tree Connect Request:
```
SMB2 Header
    Flags: 0x00000008
    Tree Id: 0x00000000
Tree Connect Request (0x03)
    StructureSize (0x0009)
    Flags: 0x0000
    Tree: \\<dfs>\<share>
```
2. Tree Connect Response, Error: STATUS_BAD_NETWORK_NAME
```
SMB2 Header:
    NT Status: STATUS_BAD_NEWTORK_NAME (0xc00000cc)
    Command: Tree Connect (3)
    Flags: 0x00000009
    Tree Id: 0x00000000
Tree Connect Response (0x03)
    StructureSize: 0x0009
    Error Data: 5c
```
3. Ioctl Request FSCTL_DFS_GET_REFERRALS, File: \<dfs>\<share>
```
SMB2 Header:
    Command: Ioctl (11)
    Flags: 0x00000008
    Tree Id: 0x00000001 \\<dfs>\IPC$
Ioctl Request (0x0b)
    Function: FSCTL_DFS_GET_REFERRALS (0x00060194)
    Flags: 0x00000001
```
4. Ioctl Response FSCTL_DFS_GET_REFERRALS
```
SMB2 Header:
    NT Status: STATUS_SUCCESS (0x00000000)
    Command: Ioctl (11)
    Flags: 0x00000009
    Tree Id: 0x00000001 \\<dfs>\IPC$
Ioctl Response (0x0b)
    Function: FSCTL_DFS_GET_REFERRALS (0x00060194)
    GUID handle:  File Id: ffffffff-ffff-ffff-ffff-fffffffffff
    Flags: 0x00000000
    In Data: NO DATA
    Out Data:
       Path Consumed: 40
       Num Referrals: 11
       Flags: 0x0003
       Referral:
         <Many with different nodes but same path>
```

**Comment by jborean93** [comment]
Nothing is jumping out here as the current workflow seems to be standard. My suspicions are with the TimeToLive (TTL) of the referrals are expired and `smbclient` isn't handling that properly. Could you share what the TTL value is of the DFS referrals returned for `<dfs><share>`?

**Comment by i3uiw** [comment]
IPv4 TTLs at 123 and 128.
For the referrals:
- Manual TTL check (ping) on some of the referral nodes in FSCTL_DFS_GET_REFERRALS.  All at or above 120.  Some packet loss between pings.
- Direct links to the nodes w/ \\<server>\<share>\<folder>\<file>.  All have valid paths.

**Comment by jborean93** [comment]
Seems like an odd number to have the TTL set at, just wanting to double check you are getting the right number? The one I'm talking about is the `TTL` in each referral response which in my example is 300, meaning the referral is meant to last 300s/5m.

![image](https://user-images.githubusercontent.com/8462645/108260531-28577300-71ae-11eb-99e2-65ed579937e6.png)

> Interesting thing to note (may be related). On a good connection, the STATUS_BAD_NETWORK_NAME still exists when connecting to \\<dfs>\<share> (Tree Connect Request). It look like code succeeds after it attempts to connect to IPC$. For the times there are fails, I do not see the additional request for IPC$.

That's how the code is structured, before we connect to a share we have no idea if it's a DFS path or just a normal server share. We attempt to connect to `\\server\share` with a Tree Connect and if that fails with `STATUS_BAD_NETWORK_NAME` we fallback to the `IPC$` share and sending the referral request.

https://github.com/jborean93/smbprotocol/blob/b72b4c7d1544b28c0d6f4b30b108d21650292be7/smbclient/_pool.py#L300-L312

The fact that you don't see the subsequent tree connect request for `\\<dfs>\IPC$` but an immediate referral is because we cache the tree connects. Once you've connected to it once then it will just be reused instead of connecting and disconnecting as required.

What I don't understand is how you are getting the `STATUS_BAD_NETWORK_NAME` when connecting to a tree. The code literally catches the `STATUS_BAD_NETWORK_NAME` exception so the only case where it would appear is if it's wrapped in an exception from `dfs_request`. I can confirm that's the case when connecting to a DFS path that doesn't exist

```
Traceback (most recent call last):
  File "/home/jborean/dev/smbprotocol/smbclient/_pool.py", line 302, in get_smb_tree
    tree.connect(require_secure_negotiate=client_config.require_secure_negotiate)
  File "/home/jborean/dev/smbprotocol/smbprotocol/tree.py", line 245, in connect
    response = self.session.connection.receive(request)
  File "/home/jborean/dev/smbprotocol/smbprotocol/connection.py", line 928, in receive
    raise SMBResponseException(response)
smbprotocol.exceptions.BadNetworkName: Received unexpected status from the server: The specified share name cannot be found on the remote server. (3221225676) STATUS_BAD_NETWORK_NAME: 0xc00000cc

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/jborean/dev/smbprotocol/jordan.py", line 24, in <module>
    print(smbclient.listdir(r'\\server2019.domain.test\bad'))
  File "/home/jborean/dev/smbprotocol/smbclient/_os.py", line 232, in listdir
    with SMBDirectoryIO(path, mode='r', share_access='r', **kwargs) as dir_fd:
  File "/home/jborean/dev/smbprotocol/smbclient/_io.py", line 291, in __init__
    tree, fd_path = get_smb_tree(path, **kwargs)
  File "/home/jborean/dev/smbprotocol/smbclient/_pool.py", line 310, in get_smb_tree
    referral = dfs_request(ipc_tree, "\\%s\\%s" % (path_split[0], path_split[1]))
  File "/home/jborean/dev/smbprotocol/smbclient/_pool.py", line 195, in dfs_request
    response = tree.session.connection.receive(request)
  File "/home/jborean/dev/smbprotocol/smbprotocol/connection.py", line 928, in receive
    raise SMBResponseException(response)
smbprotocol.exceptions.NotFound: Received unexpected status from the server: The object was not found. (3221226021) STATUS_NOT_FOUND: 0xc0000225
```

You can see the original exception was `BadNetworkName` and when we tried to send the dfs referral request it then failed with `NotFound`. Are you sure you got the full exception information?

**Comment by i3uiw** [comment]
Ahh!  Wrong TTL!  My bad. Referral TTLs are all 300.

First referral:

```
Referral
  Version: 4
  Size: 34
  Server Type: Root targets returns (1)
    Flags: 0x0004, TargetSetBoundary
      .... .... .... ..0. = NameListReferral: NOT a domain/DC referral response
      .... .... .... .1.. = TargetSetBoundary: The first target in the target set
  TTL: 300
  Path Offset: 374
  Alt Path Offset: 416
  Node Offset: 458
  Server GUID: 00000000-0000-0000-0000-000000000000
  Path: \<dfs>\<share>
  Alt Path: \<dfs>\<share>
  Node: \<server>\<share>

```
I can confirm that is the full trace back unless you want to see the max recursion details.
Like you, I'm just as confused as well as everything we are seeing should be working.

**Comment by jborean93** [comment]
So what's the last Tree Connect you have in your wireshark captured that returned `STATUS_BAD_NETWORK_NAME`. The only time I can see `BadNetworkName` being raised is if you tried to connect to `\\server\IPC$` itself and it raised that.

**Comment by i3uiw** [comment]
Wireshark shows multiple `STATUS_BAD_NETWORK_NAME` events as Python is also showing the events.  Until Python hits its max recursion.

I tried moving the DFS and share names to lower case w/ the same results (previously was a mix of lower and upper).  Folder paths were left as is (mix of upper/lower case) since `STATUS_BAD_NETWORK_NAME` indicates a problem with getting IP/DNS resolved.

**Comment by jborean93** [comment]
For it to hit the max recursion it’s bound to send multiple requests to the same host, are you not seeing that?

**Comment by i3uiw** [comment]
Yes I am.

**Comment by jborean93** [comment]
So what's the share it's trying to connect to, does it match `\\<dfs>\IPC$` exactly or does the case differ?

**Comment by i3uiw** [comment]
No references to IPC$ are being made on the recursion failure.  Its like the python code is using the cache but the cache is bad somehow.  So in the failure situation it is like this:
```

1. Tree Connect Request \\<dfs>\<share>
2. Tree Connect Request Error: STATUS_BAD_NETWORK_NAME
3. Ioctl Request FSCTL_DFS_GET_REFERRALS
4. Ioctl Response FSCTL_DFS_GET_REFERRALS
5. Tree Connect Request \\<dfs>\<share>
6. Tree Connect Request Error: STATUS_BAD_NETWORK_NAME
7. Ioctl Request FSCTL_DFS_GET_REFERRALS
8. Ioctl Response FSCTL_DFS_GET_REFERRALS
9. Tree Connect Request \\<dfs>\<share>
10. Tree Connect Request Error: STATUS_BAD_NETWORK_NAME
11. Ioctl Request FSCTL_DFS_GET_REFERRALS
12. Ioctl Response FSCTL_DFS_GET_REFERRALS
13. Repeats until python hits max recursion
```

**Comment by jborean93** [comment]
Ok, so in the `FSCTL_DFS_GET_REFERRALS` response (step 3) you have a list of referrals for `\\<dfs>\<share>` with the `Path` being the DFS path and `Node` being the target path. Does the `Path` in all the referrals match exactly with `\\<dfs>\<share>` or does the case differ in anyway? It sounds like we are sending the referral request, caching the response but when we go to lookup the referral cache we are unable to find any hints because our paths components aren't matching up.

**Comment by i3uiw** [comment]
Step 4, 8, 12, etc:
Each referral has the same path of `\<dfs>\<share>` and alt path of '\<dfs>\<share>`.  They differ in node with `\<node>\<share>`.  These `\<node>\<share>` are the ones I checked to confirm are functioning and available (at a minimum).

Between each FSCTL_DFS_GET_REFERRALS , the list of referrals change in order only.  Same list of 5+ referrals.

**Comment by jborean93** [comment]
Is the casing any different, like I'm connecting to `\\DFS\share` and the returned DFS path in the referral is `\dfs\Share` or `\DFS\share` or some other unique combination.

**Comment by i3uiw** [comment]
Casing all matches for all `<dfs>\share`.  When switching to node, `<node>` is in uppercase (<share> has not changed cases).  Not FQDN.

**Comment by jborean93** [comment]
I'm sorry at this point I don't know what is the problem and because I cannot replicate it there's not much more I can do. You are going to have to step through [get_smb_tree](https://github.com/jborean93/smbprotocol/blob/b72b4c7d1544b28c0d6f4b30b108d21650292be7/smbclient/_pool.py#L225) with a debugger yourself, in particular https://github.com/jborean93/smbprotocol/blob/b72b4c7d1544b28c0d6f4b30b108d21650292be7/smbclient/_pool.py#L300-L312 to figure out why it's not matching the path with the cached referrals and constantly looping.

What should be happening is it

* sends the `FSCTL_DFS_GET_REFERRALS` for `\<dfs>\<share>`
* this returns 1 or more referrals for `\<dfs>\<share>` that point to `\<server>\<share>`
* it then tries to lookup the referral cache for `\\<dfs>\<share>` that corresponds with the DFS path returned in each referral
* once a match is found it replaces the path `\\<dfs>\<share>` with the node path in that matching referral
* it then tries to connect to that replaced path and repeats the process

Somewhere in here, I assume the referral lookup isn't finding a match when it should, is not working properly and you are going to have to narrow it down as based on this conversation I've exhausted all the things I can think off.

I'm sure there is a problem somewhere but unfortunately I cannot fix what I don't know is broken.

**Comment by i3uiw** [comment]
Much appreciated for the help!  This is an annoying one.  If I find a solution, I'll report back.

**Comment by i3uiw** [comment]
UPDATE!
TLDR: I didn't find the issue but worked around it.  Suspect network issues.
Changes were made to my script:
1. Reset cache before every write via `smbclient.reset_connection_cache()`
2. Wrote data into memory via `io.BytesIO()` file-like object, compressed via zipfile module in memory and wrote the compressed zip once.

Thanks for the assist!  I greatly appreciate it!

**Comment by jborean93** [comment]
No worries, glad you found a workaround. If you do come across it again or find more information feel free to share whatever you see.
